### ISSUE TITLE
Save img to context

### ISSUE THREAD

**Issue by greyser83** [bartbutenaers/node-red-contrib-ui-heatmap#5]
Bart, sorry to bother.
Is there any solution to grab heatmap image? And somehow save it to context

**Comment by bartbutenaers** [comment]
Hi @greyser83,

That is again very easy to implement.   When I add an output to this node, and I add following code to my .js file:
```
$scope.$watch('msg', function(msg) {
   ...
   // Get a reference to the heatmap.js canvas element
   var heatmapCanvas = parentDiv.getElementsByClassName("heatmap-canvas")[0];
                                
   // Get the canvas as a jpeg image inside a data url (data:image/jpeg;base64,/9j/4AAQSkZJRgABAQ...9oADAMBAAIRAxEAPwD/AD/6AP/Z")
   var dataUrl = heatmapCanvas.toDataURL('image/jpeg', 1.0); // 1.0 = full quality

   // Get the base64 image from the data url
   var base64Image = dataUrl.split(';base64,')[1];

   $scope.send({payload: base64Image});
});
```
Then a jpeg will be send to the output, for every input message that arrives:

![heatmap_image](https://user-images.githubusercontent.com/14224149/109710987-a3515e00-7b9e-11eb-9467-a925f48c31d0.gif)

In the above example the _**backend triggers the capture**_: But it is not clear to me how this would be used.  Because the image is generated at the frontend side (i.e. in the dashboard running in the browser) and then send as output message to the backend (where your flow is running).  As a result, when the heatmap is currently visible in N dashboards, then N output messages will be send.  So you end up with N duplicate messages (each containing the same image).  And N can be 0: when no dashboards are open currently, then 0 messages would be send.

To me this only makes sense when _**frontend triggers the capture**_.  For example you add a button _"create heatmap image"_ on your dashboard.  As soon as a user triggers that button, the button will send a message to the backend flow, where it is injected into the heatmap node (e.g. with topic "take_snapshot").  The heatmap node will be send by the heatmap node to the frontend to take a snapshot.  

So would be nice to know your use case, before I add this code on my master branch.

**Comment by greyser83** [comment]
Hi, Bart. Thank you so much for your reply, very impressive.My task is to send a heatmap image to telegram bot. 23:45, 2 марта 2021 г., bartbutenaers <notifications@github.com>:
Hi @greyser83,
That is again very easy to implement.   When I add an output to this node, and I add following code to my .js file:
$scope.$watch('msg', function(msg) {
   ...
   // Get a reference to the heatmap.js canvas element
   var heatmapCanvas = parentDiv.getElementsByClassName("heatmap-canvas")[0];
                                
   // Get the canvas as a jpeg image inside a data url (data:image/jpeg;base64,/9j/4AAQSkZJRgABAQ...9oADAMBAAIRAxEAPwD/AD/6AP/Z")
   var dataUrl = heatmapCanvas.toDataURL('image/jpeg', 1.0); // 1.0 = full quality

   // Get the base64 image from the data url
   var base64Image = dataUrl.split(';base64,')[1];

   $scope.send({payload: base64Image});
});

Then a jpeg will be send to the output, for every input message that arrives:

In the above example the backend triggers the capture: But it is not clear to me how this would be used.  Because the image is generated at the frontend side (i.e. in the dashboard running in the browser) and then send as output message to the backend (where your flow is running).  As a result, when the heatmap is currently visible in N dashboards, then N output messages will be send.  So you end up with N duplicate messages (each containing the same image).  And N can be 0: when no dashboards are open currently, then 0 messages would be send.
To me this only makes sense when frontend triggers the capture.  For example you add a button "create heatmap image" on your dashboard.  As soon as a user triggers that button, the button will send a message to the backend flow, where it is injected into the heatmap node (e.g. with topic "take_snapshot").  The heatmap node will be send by the heatmap node to the frontend to take a snapshot.
So would be nice to know your use case, before I add this code on my master branch.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.
-- Отправлено из мобильного приложения Яндекс.Почты

**Comment by bartbutenaers** [comment]
> My task is to send a heatmap image to telegram bot

Hi @greyser83 ,
That is not really much information for me to understand what you are doing ...
How do you trigger the image generation: via a user action (e.g. button click) in the dashboard, or in background automatically by some other nodes in your flow that trigger an input message?
Because in the latter case my solution won't work at all, since you can have N dashboards open (where N can be 0): see my explanation above.

**Comment by greyser83** [comment]
Hi @bartbutenaers 
> How do you trigger the image generation: via a user action (e.g. button click) in the dashboard, or in background automatically by some other nodes in your flow that trigger an input message?
> Because in the latter case my solution won't work at all, since you can have N dashboards open (where N can be 0): see my explanation above.

Image generation must be triggerd by bot (user presses a button), which sends a message to NR. Image goes back to bot as a response
![image](https://user-images.githubusercontent.com/49680277/109805212-3688a300-7c34-11eb-960c-56d0d88c384d.png)
![image](https://user-images.githubusercontent.com/49680277/109805303-55873500-7c34-11eb-98c7-3ca0ffee94db.png)

**Comment by bartbutenaers** [comment]
Hi @greyser83,

I have implemented your request on Github.  There is also a new [section ](https://github.com/bartbutenaers/node-red-contrib-ui-heatmap/blob/master/README.md#snapshot)on the readme page which explains it.

Would be nice if you could test it and give feedback!
Note that I need a surgery this friday, and I won't be able to do developments for a few months...
So not sure whether we will get this published on NPM tomorrow.  And otherwise you will need to use the Github version for some time ...

Bart

**Comment by greyser83** [comment]
Hi, Bart!
I tried to install newest version and it is installed as I see in logs. But I cant see new node in palette
![image](https://user-images.githubusercontent.com/49680277/109923972-a2204e00-7cd0-11eb-978b-b62205c183c4.png)
![image](https://user-images.githubusercontent.com/49680277/109924008-afd5d380-7cd0-11eb-8757-2c330bc09b82.png)
![image](https://user-images.githubusercontent.com/49680277/109924035-bc5a2c00-7cd0-11eb-9653-c024c4c4686f.png)

But anyway its not urgent to me! Many thanks!

I wish you a quick, easy recovery!

**Comment by bartbutenaers** [comment]
The command looks good, so not clear to me why it doesn't work.
You can have a look at your browser console log (via the browser development tools), to see if there is an error.  Or in your Node-RED log on linux.  And always refresh your browser window (where the flow editor is running).
My time is up.  I will bookmark this issue, and come back to you in the near future...

**Comment by bartbutenaers** [comment]
Hi @greyser83,
Long time ago ...
Did you see any errors in your browser console log?
I have installed it on another system and it works fine.
Bart

**Comment by greyser83** [comment]
Hi, Bart!I successfully installed and it works precisely since when. Really appreciate your stuff and pleased to see you again! Thanks!22.05.2021, 01:17, "bartbutenaers" ***@***.***>:
Hi @greyser83,
Long time ago ...
Did you see any errors in your browser console log?
I have installed it on another system and it works fine.
Bart

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.
-- С уважением,Сергей Костромин

**Comment by bartbutenaers** [comment]
Hi @greyser83,
This feature has been published on npm now, in version ***2.2.0***.
Thanks for testing!!
Bart
